lib/OpenQA/WebAPI/Plugin/AMQP.pm

```diff
@@ -70,6 +71,179 @@ sub log_event {
+            id   => $artifact_id,
+        },
+        pipeline => {
+            # per https://pagure.io/fedora-ci/messages/issue/61 this
+            # is meant to be unique per test scenario *and* artifact,
+            # so we construct it out of BUILD and the scenario keys.
+            # 'name' is supposed to be a 'human readable name', well,
+            # this is human readable, so we'll just use it twice
+            id   => join('.', "openqa", $job->BUILD, $job->TEST, $job->MACHINE, $job->FLAVOR, $job->ARCH),
+            name => join('.', "openqa", $job->BUILD, $job->TEST, $job->MACHINE, $job->FLAVOR, $job->ARCH),
+        },
+        test => {
+            # openQA tests are pretty much always validation
+            category => 'validation',
+            # test identifier: test name plus scenario keys
+            type      => join(' ', $job->TEST, $job->MACHINE, $job->FLAVOR, $job->ARCH),
+            namespace => $test_namespace,
+        },
+        system => {
+            # it's interesting whether we should record info on the
+            # *worker host itself* or the *SUT* (the VM run on top of
+            # the worker host environment) here...on the whole I think
+            # SUT is more in line with expectations, so let's do that
+            os => "fedora-${sysrelease}",
+            # openqa provisions itself...we *could* I guess set this
+            # to 'createhdds' if we booted a disk image, but ehhhh
+            provider     => 'openqa',
+            architecture => $job->ARCH,
+            variant      => $job->settings_hash->{SUBVARIANT},
+        },
+        generated_at => $generated_at,
+        version      => "0.2.1",
+    );
+
+    # add keys that don't exist in all cases to the message
+    if ($stdevent eq 'complete') {
+        $msg_data{test}{result} = $job->result;
+        $msg_data{test}{result} = 'info' if $job->result eq 'softfailed';
+    }
+    elsif ($stdevent eq 'error') {
+        $msg_data{error} = {};
+        $msg_data{error}{reason} = $job->result;
+    }
+    elsif ($stdevent eq 'queued') {
+        # this is a hint to consumers that the job probably went away
+        # if they don't get a 'complete' or 'error' in 4 hours
+        # FIXME: we should set this as 2 hours on 'running', but we
+        # can't emit running ATM...
+        $msg_data{test}{lifetime} = 240;
+    }
+    $msg_data{run}{clone_of} = $clone_of if ($clone_of);
+
+    $msg_data{artifact}{release} = $artifact_release if ($artifact_release);
+
+    $msg_data{artifact}{compose_type} = $compose_type if ($compose_type);
+
+    $msg_data{artifact}{iso} = $job->settings_hash->{ISO} if ($job->settings_hash->{ISO});
+    # FIXME: hdd_2?
+    $msg_data{artifact}{hdd_1} = $job->settings_hash->{HDD_1} if ($job->settings_hash->{HDD_1});
+
+    # convert data to JSON, with reliable key ordering (helps the tests)
+    my $msg_json = Cpanel::JSON::XS->new->canonical(1)->allow_blessed(1)->encode(\%msg_data);
```

Comment:
The default `to_json` or `encode_json` from `Mojo::JSON` wouldn't work here?

Reply:
I didn't even write this. It is already there. :P

Reply:
looking at the history, it seems you converted a lot of `Cpanel::JSON` uses to `to_json` in 2bc55edf , but then converted some back to `Cpanel::JSON` again in 4d8cf50a . The use in `Fedmsg.pm` was one of these. You didn't change the use in `AMQP.pm`'s existing `log_event` in either of those commits, you left it as `Cpanel::JSON`.

Reply:
Right, i remember that. I wasn't sure about protocol requirements back then. Avoiding string comparisons for JSON when they are not absolutely necessary sounds like a very good idea though.

Reply:
so yeah, from the docs the `Mojo::JSON` functions don't seem to have any ability to do key ordering, so we can't change the use in `log_event` unless we also change the tests as discussed. We could just use `to_json` for `log_event_fedora_ci_messages` I guess, but it kinda makes sense to me for them both to use the same thing...so if we're gonna change this maybe do it as a follow-up that changes `23-amqp.t` and then changes both of these?

Reply:
Key ordering in `Mojo::JSON` is the default and that's a feature, but i guess it's missing from the docs.

Reply:
ah OK, I guess I can try switching them both and see if `23-amqp.t` still passes...

Reply:
OK, so I tried using `to_json` instead...it basically works, only it escapes URLs where `Cpanel::Json::XS` does not. This makes `23-amqp.t` fail, and more to the point, it *would* be a change in the actual format of the messages. I'm not really sure how big of a deal it is, but it at least seems like maybe something to consider separately from this PR, so for now I think I'll leave both using Cpanel.